presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoinsUtils.java

```diff
@@ -0,0 +1,183 @@
+     * In order not to generate the inverse of any set, we always include the 0th
+     * node in our sets.
+     *
+     * @param totalNodes
+     * @return A set of sets each of which defines a partitioning of totalNodes
+     */
+    public static Set<Set<Integer>> generatePartitions(int totalNodes)
+    {
+        return generatePartitions(ImmutableSet.of(0), totalNodes, 1);
+    }
+
+    private static Set<Set<Integer>> generatePartitions(Set<Integer> existingSet, int totalNodes, int index)
+    {
+        ImmutableSet.Builder<Set<Integer>> sets = ImmutableSet.builder();
+        if (index < totalNodes && existingSet.size() < totalNodes - 1) {
+            ImmutableSet.Builder<Integer> setWithAddedNode = ImmutableSet.builder();
+            setWithAddedNode.addAll(existingSet);
+            setWithAddedNode.add(index);
+            sets.addAll(generatePartitions(setWithAddedNode.build(), totalNodes, index + 1));
+            sets.addAll(generatePartitions(existingSet, totalNodes, index + 1));
+        }
+        else {
+            sets.add(existingSet);
+        }
+        return sets.build();
+    }
+
+    public static JoinNode createJoinAccordingToPartitioning(JoinGraphNode joinGraph, Set<Integer> leftSources, PlanNodeIdAllocator idAllocator)
+    {
+        List<PlanNode> sources = joinGraph.getSources();
+        ImmutableList.Builder<PlanNode> leftNodes = ImmutableList.builder();
+        List<Symbol> leftSymbols = new ArrayList<>();
+        ImmutableList.Builder<PlanNode> rightNodes = ImmutableList.builder();
+        List<Symbol> rightSymbols = new ArrayList<>();
+        for (int i = 0; i < sources.size(); i++) {
+            PlanNode source = sources.get(i);
+            if (leftSources.contains(i)) {
+                leftNodes.add(source);
+                leftSymbols.addAll(source.getOutputSymbols());
+            }
+            else {
+                rightNodes.add(source);
+                rightSymbols.addAll(source.getOutputSymbols());
+            }
+        }
+
+        ImmutableList.Builder<JoinNode.EquiJoinClause> leftCriteria = ImmutableList.builder();
+        ImmutableList.Builder<JoinNode.EquiJoinClause> rightCriteria = ImmutableList.builder();
+        ImmutableList.Builder<JoinNode.EquiJoinClause> spanningCriteria = ImmutableList.builder();
+        for (JoinNode.EquiJoinClause equiJoinClause : joinGraph.getCriteria()) {
+            Set<Symbol> equiJoinClauseSymbols = ImmutableSet.of(equiJoinClause.getLeft(), equiJoinClause.getRight());
+            if (leftSymbols.containsAll(equiJoinClauseSymbols)) {
+                leftCriteria.add(equiJoinClause);
+            }
+            else if (rightSymbols.containsAll(equiJoinClauseSymbols)) {
+                rightCriteria.add(equiJoinClause);
+            }
+            else {
+                spanningCriteria.add(equiJoinClause);
+            }
+        }
+
+        ImmutableList.Builder<Expression> leftFilters = ImmutableList.builder();
```

Comment:
We though about it with @losipiuk and it seems that you should use `EqualityInference` for selecting which filters and conditions can be used in what subtree (See `com.facebook.presto.sql.planner.optimizations.PredicatePushDown.Rewriter#processInnerJoin` how this can be done). When you use equality inference you will implicitly pushdown all filters to appropriate join nodes and correctly compute cost for subtree. Without equality inference you might end up with cross joins for which cost would explode.